contractsSol5/KyberHintHandler.sol

```diff
@@ -346,22 +177,139 @@ contract KyberHintHandler is IKyberHint, Utils4 {
+            TradeType e2tType,
+            IKyberReserve[] memory e2tAddresses,
+            uint[] memory e2tSplits,
+            HintErrors verified
+        )
+    {
+        bytes8[] memory t2eReserveIds;
+        bytes8[] memory e2tReserveIds;
+
+        (
+            t2eType,
+            t2eReserveIds,
+            t2eSplits,
+            e2tType,
+            e2tReserveIds,
+            e2tSplits
+        ) = abi.decode(hint, (TradeType, bytes8[], uint[], TradeType, bytes8[], uint[]));
+        HintErrors verifiedT2E = verifyData(t2eType, t2eReserveIds, t2eSplits);
+        HintErrors verifiedE2T = verifyData(e2tType, e2tReserveIds, e2tSplits);
+
+        if (verifiedT2E == HintErrors.NoError && verifiedE2T == HintErrors.NoError) {
+            t2eAddresses = new IKyberReserve[](t2eReserveIds.length);
+            e2tAddresses = new IKyberReserve[](e2tReserveIds.length);
+            verified = HintErrors.NoError;
+            
+            for (uint i = 0; i < t2eReserveIds.length; i++) {
+                t2eAddresses[i] = IKyberReserve(convertReserveIdToAddress(t2eReserveIds[i]));
             }
-            for (uint i = 0; i < ethToTokenAddresses.length; i++) {
-                ethToTokenReserveIds[i] = convertAddressToReserveId(address(ethToTokenAddresses[i]));
+            for (uint i = 0; i < e2tReserveIds.length; i++) {
+                e2tAddresses[i] = IKyberReserve(convertReserveIdToAddress(e2tReserveIds[i]));
             }
+        } else {
+            t2eSplits = new uint[](0);
+            e2tSplits = new uint[](0);
+
+            if (verifiedT2E != HintErrors.NoError) verified = verifiedT2E;
+            if (verifiedE2T != HintErrors.NoError) verified = verifiedE2T;
         }
     }
 
+    /// @notice Ensures that the data passed when building/parsing hints is valid
+    /// @param tradeType Trade hint type
+    /// @param reserveIds Reserve IDs
+    /// @param splits Reserve splits
+    /// @return returns a HintError enum to indicate valid or invalid hint data
+    function verifyData(
+        TradeType tradeType,
+        bytes8[] memory reserveIds,
+        uint[] memory splits
+    )
+        internal
+        pure
+        returns (HintErrors)
+    {
+        if (!(reserveIds.length > 0)) return HintErrors.ReserveIdZeroError;
+        if (tradeType == TradeType.Split) {
+            if (reserveIds.length != splits.length) return HintErrors.ReserveIdSplitsError;
+
+            uint bpsSoFar;
+            for (uint i = 0; i < splits.length; i++) {
+                bpsSoFar += splits[i];
+            }
+
+            if (bpsSoFar != BPS) return HintErrors.TotalBPSError;
+        } else {
+            if (splits.length != 0) return HintErrors.SplitsZeroError;
+        }
+
+        return HintErrors.NoError;
+    }
+
+    /// @notice Throws error message to user to indicate error on hint
+    /// @param error Error type from HintErrors enum
+    function throwHintError(HintErrors error) internal pure {
```

Comment:
please try a bit shorter message
this is still part of matching engine.
so should be readable
but shorter.

Reply:
Okay, will try to shorten messages.